tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc

```diff
@@ -558,6 +560,94 @@ static std::vector<string> GetROCDLPaths(int amdgpu_version,
+// TargetMachine for the AMDGPU target.
+StatusOr<std::vector<uint8>> EmitModuleToHsaco(
+    Module* module, llvm::TargetMachine* target_machine) {
+  char tempdir_template[] = "/tmp/amdgpu_xla-XXXXXX";
+  char* tempdir_name = mkdtemp(tempdir_template);
+
+  VLOG(1) << "Compile-time artifacts located at: " << tempdir_name;
+
+  // prepare filenames for all stages of compilation:
+  // IR, ISA, binary ISA, and HSACO
+  std::string ir_filename = absl::StrCat(module->getModuleIdentifier(), ".ll");
+  std::string ir_path = tensorflow::io::JoinPath(tempdir_name, ir_filename);
+
+  std::string isabin_filename =
+      absl::StrCat(module->getModuleIdentifier(), ".o");
+  std::string isabin_path =
+      tensorflow::io::JoinPath(tempdir_name, isabin_filename);
+
+  std::string hsaco_filename =
+      absl::StrCat(module->getModuleIdentifier(), ".hsaco");
+  std::string hsaco_path =
+      tensorflow::io::JoinPath(tempdir_name, hsaco_filename);
+
+  std::error_code ec;
+
+  // dump LLVM IR
+  std::unique_ptr<llvm::raw_fd_ostream> ir_fs(
+      new llvm::raw_fd_ostream(ir_path, ec, llvm::sys::fs::F_None));
+  module->print(*ir_fs, nullptr);
+  ir_fs->flush();
+
+  //// emit GCN ISA binary
+  // The extension is stripped by IrDumpingPassManager, so we need to
+  // get creative to add a suffix.
+  std::string module_id = module->getModuleIdentifier();
+  IrDumpingPassManager codegen_passes(
+      ReplaceFilenameExtension(tensorflow::io::Basename(module_id),
+                               "-amdgpu.dummy"),
+      "", false);
+  codegen_passes.add(new llvm::TargetLibraryInfoWrapperPass(
+      llvm::Triple(module->getTargetTriple())));
+  llvm::SmallVector<char, 0> stream;
+  llvm::raw_svector_ostream pstream(stream);
+  std::unique_ptr<llvm::raw_fd_ostream> isabin_fs(
+      new llvm::raw_fd_ostream(isabin_path, ec, llvm::sys::fs::F_Text));
+  module->setDataLayout(target_machine->createDataLayout());
+  target_machine->addPassesToEmitFile(codegen_passes, *isabin_fs, nullptr,
+                                      llvm::TargetMachine::CGFT_ObjectFile);
+  codegen_passes.run(*module);
+  isabin_fs->flush();
+
+  // Locate lld
+  // ROCM TODO: change to tensorflow::ROCmRoot() after ROCm-Device-Libs PR.
+  std::string lld_path = tensorflow::io::JoinPath("/opt/rocm", "hcc/bin");
+  auto lld_program = llvm::sys::findProgramByName("ld.lld", {lld_path});
+  if (!lld_program) {
+    LOG(FATAL) << "unable to find ld.lld in PATH: "
+               << lld_program.getError().message();
+  }
+  std::vector<llvm::StringRef> lld_args{
+      llvm_ir::AsStringRef("ld.lld"),      llvm_ir::AsStringRef("-flavor"),
+      llvm_ir::AsStringRef("gnu"),         llvm_ir::AsStringRef("-shared"),
+      llvm_ir::AsStringRef("isabin_path"), llvm_ir::AsStringRef("-o"),
+      llvm_ir::AsStringRef("hsaco_path"),
+  };
+  lld_args[4] = llvm_ir::AsStringRef(isabin_path.c_str());
+  lld_args[6] = llvm_ir::AsStringRef(hsaco_path.c_str());
```

Comment:
Please initialize lld_args[4] and [6] in place.

Reply:
will revise.